tools/launch.py

```diff
@@ -0,0 +1,135 @@
+"""Launching tool for DGL distributed training"""
+import os
+import sys
+import subprocess
+import argparse
+import signal
+import logging
+import time
+from threading import Thread
+
+def execute_remote(cmd, ip):
+    """execute command line on remote machine via ssh"""
+    cmd = 'ssh -o StrictHostKeyChecking=no ' + ip + ' \'' + cmd + '\''
+    # thread func to run the job
+    def run(cmd):
+        subprocess.check_call(cmd, shell = True)
+
+    thread = Thread(target = run, args=(cmd,))
+    thread.setDaemon(True)
+    thread.start()
+
+def kill_python_proc(args):
+    """Kill all the python processes accross machines via ssh"""
+    ip_config = args.workspace + '/' + args.ip_config
+    with open(ip_config) as f:
+        for line in f:
+            ip, _, _ = line.strip().split(' ')
+            cmd = 'pkill -9 python'
+            execute_remote(cmd, ip)
+
+    # use ctl+C to exit
+    while True:
+        time.sleep(10)
+
+def submit_jobs(server_cmd, client_cmd, args):
+    """Submit distributed jobs (server and client processes) via ssh"""
+    hosts = []
+    server_count_per_machine = 0
+    ip_config = args.workspace + '/' + args.ip_config
+    with open(ip_config) as f:
+        for line in f:
+            ip, port, count = line.strip().split(' ')
+            port = int(port)
+            count = int(count)
+            server_count_per_machine = count
+            hosts.append((ip, port))
+    assert args.num_client % len(hosts) == 0
+    client_count_per_machine = args.num_client / len(hosts)
+    # launch server tasks
+    for i in range(len(hosts)*server_count_per_machine):
+        ip, _ = hosts[int(i / server_count_per_machine)]
+        cmd = server_cmd + ' --id ' + str(i)
+        cmd = 'cd ' + str(args.workspace) + '; ' + cmd
+        execute_remote(cmd, ip)
+    # launch client tasks
+    for i in range(args.num_client):
+        ip, _ = hosts[int(i / client_count_per_machine)]
+        cmd = client_cmd.replace('node_rank=0', 'node_rank='+str(i))
+        cmd = 'cd ' + str(args.workspace) + '; ' + cmd
+        execute_remote(cmd, ip)
+
+    # use ctl+C to exit
+    while True:
+        time.sleep(10)
+
+def server_command(args):
+    server_cmd = 'python3 ' + str(args.exe_file) + ' --server'
+    server_cmd = server_cmd + ' --graph-name ' + str(args.graph_name)
+    server_cmd = server_cmd + ' --num-client ' + str(args.num_client)
+    server_cmd = server_cmd + ' --conf_path ' + str(args.conf_path)
+    server_cmd = server_cmd + ' --ip_config ' + str(args.ip_config)
+    return server_cmd
```

Comment:
we shouldn't ask the users to write code for servers. we can set dgl into a server mode and run services automatically.
mxnet starts service like this. you can use environment variables to set dgl into the server mode.

Reply:
Users will not access the server code. This is just for our scheduler.